packages/navie/src/user-context.ts

```diff
@@ -0,0 +1,75 @@
+export namespace UserContext {
+  export type CodeSelectionItem = {
+    type: 'code-selection';
+    content: string;
+  };
+
+  export type LocationItem<T = any> = {
+    type: T;
+    location?: string;
+  };
+
+  export type CodeSnippetItem = LocationItem<'code-snippet'> & {
+    content: string;
+  };
+
+  export type FileItem = LocationItem<'file'>;
+
+  // A specific item provided by the user
+  export type ContextItem = CodeSelectionItem | CodeSnippetItem | FileItem;
+  export type Context = string | ContextItem[];
+
+  export function isCodeSelectionItem(i: ContextItem): i is CodeSelectionItem {
+    return i.type === 'code-selection';
+  }
+
+  export function isCodeSnippetItem(i: ContextItem): i is CodeSnippetItem {
+    return i.type === 'code-snippet';
+  }
+
+  export function isFileItem(i: ContextItem): i is FileItem {
+    return i.type === 'file';
+  }
+
+  export function hasLocation(i: ContextItem): i is LocationItem & { location: string } {
+    return 'location' in i && i.location !== undefined;
+  }
+
+  export function renderItems(items: ContextItem[], includeContent: boolean = true) {
+    const renderItem = (i: UserContext.ContextItem) => {
+      if (UserContext.isCodeSelectionItem(i) && includeContent) {
+        return `<code-selection>\n${i.content}\n</code-selection>`;
+      } else if (UserContext.isCodeSnippetItem(i)) {
+        const content = includeContent ? `\n<content>${i.content}</content>\n'` : '';
+        return `<pinned-snippet>
+  <location>${i.location}</location>${content}
+</pinned-snippet>`;
+      } else if (UserContext.isFileItem(i)) {
+        return `<pinned-file><location>${i.location}</location></pinned-file>`;
```

Comment:
Pinned files never include content?

Reply:
That's right, pinned files never include content. #2049 is about stale file contents. These changes avoid that by having context selection always read the file.

Reply:
One of the significant changes here is this: https://github.com/getappmap/appmap-js/blob/explain-locations_20241007/packages/navie/src/agent.ts#L36 . When a `ContextItem` has a `location`, it's included in the `ContextFilters`. As I understand it, this is the same mechanism that `@generate` uses to read the full contents of files.

You can see this in action in Navie's log:
```
[context-service] Searching for context
Tokens (prompt/compl/total): 657/26/683, cost: $0.00
Vector terms response:
Context: Exception handling in Java
Instructions: Explain how exception handling is managed
---
Terms: exception handling +Java try catch finally
Explain received context request: search
Collecting context with parameters: sourceDirectories: /Users/ajp/src/apotterri/ejb-example, charLimit: 16503, appmapDirectories: /Users/ajp/src/apotterri/ejb-example, keywords: exception, handling, java, try, catch, finally, locations: /Users/ajp/src/apotterri/ejb-example/src/main/java/org/superbiz/remote/App.java
Parsed locations: /Users/ajp/src/apotterri/ejb-example/src/main/java/org/superbiz/remote/App.java
[location-context] Candidate locations: /Users/ajp/src/apotterri/ejb-example/src/main/java/org/superbiz/remote/App.java
[location-context] Extracting snippet for location: /Users/ajp/src/apotterri/ejb-example/src/main/java/org/superbiz/remote/App.java (2127 bytes)
```

What you see under Code Snippets is in addition to the contents of files. This redundancy is what this issue is about: https://github.com/getappmap/appmap-js/issues/1984 .